File: Java/Euler04/Euler04_b.java
/*A palindromic number reads the same both ways. 
The largest palindrome made from the product of 
two 2-digit numbers is 9009 = 91 × 99.
Find the largest palindrome made from the product 
of two 3-digit numbers. 
(Answer = 906609) */

// Use stringbuilder to reverse string
// ~ 14,000,000 nanoseconds

/**
 * @author ccnelson 2020
 */
public class Euler04_b
{
    public static void main(String[] args)
    {
        long startTime;
        long endTime;
        long duration;
        // start timer
        startTime = System.nanoTime(); 

        int temp = 0;
        int answer = 0;
        int val1 = 0;
        int val2 = 0;
        
        for (int i = 1000; i > 900; i--)
        {
            for (int j = 1000; j > 900; j--)
            {
                temp = i * j;
                if (Revnumstring(temp) == temp && temp > answer)
                {
                    answer = temp;
                    val1 = i;
                    val2 = j;
                }
            }
        }

        endTime = System.nanoTime();
        duration = (endTime - startTime);
        
        System.out.print("Answer: " + answer);
        System.out.println(" (by multiplying " + val1 + " & " + val2 + ")");
        System.out.println("Took " + duration + " nanoseconds");
    }

    public static int Revnumstring(int n)
    {
        StringBuilder stringBuild = new StringBuilder();
        stringBuild.append(Integer.toString(n));
        String revString = stringBuild.reverse().toString();
        int output = Integer.valueOf(revString);
        return output;
    }
}

File: Python/Euler04/Euler04_b.py
# A palindromic number reads the same both
# ways. The largest palindrome made from
# the product of two 2-digit numbers is
# 9009 = 91 × 99.
# Find the largest palindrome made from
# the product of two 3-digit numbers.
# (Answer = 906609)

# Uses math method to reverse positive integer
# ~ 30,000,000 nanoseconds

import time

def Revnum(n):
    reverse = 0
    endDigit = 0
    while n > 0:
        endDigit = n % 10
        reverse = (reverse * 10) + endDigit 
        n = n // 10; # floor division
    return reverse

start = time.perf_counter_ns()

answer = 0

for i in range(1000, 900, -1):
    for j in range(1000, 900, -1):

        temp = i * j
        if temp == Revnum(temp) and temp > answer:
            answer = temp
            val1 = i
            val2 = j

duration = time.perf_counter_ns() - start

print(answer)
print("Product of ", val1, " and ", val2)
print("Took: %s nanoseconds" % (duration))